### PR TITLE
Add callback and non-blocking option to StudyDeck

### PR DESCRIPTION
**Pull request by roxgib**
### Background

QDialog has a few different ways of displaying itself, including:

```
QDialog.exec() # runs in own loop, blocks program until finished
QDialog.open() # non-blocking
```

 The blocking version is useful if your function needs to get a value out of QDialog, but this causes issues because:
- If QDialog causes an error, it will also cause an error in the function that called it (or simply hang)
- Other code might need to run in the background, e.g. an add-on that makes a selection from the dialog

A similar problem exists with QMenu, and ideally all subclasses of QDialog and QMenu will be changed to their respective non-blocking options, but StudyDeck is known to be causing issues (see #987) already. However, since existing code (and likely add-ons as well) relies on QDialogs being blocking, this will cause issues if the behaviour is changed.

### Fix

A `callback` kwarg is added to StudyDeck, accepting a callable and defaulting to `None`, and the class is modified so that:
- if `callback == None` the class retains its old, blocking, behaviour. As this is the default, existing code that calls StudyDeck will be unaffected. 
- if a callable is passed to callback, the callable is called with the StudyDeck instance as its only argument by `StudyDeck.accept`. New code that requires the result of StudyDeck should use this behaviour.

This option allows existing code (and addons) to be updated while preserving legacy behaviour in the interim. Once the new behaviour is confirmed to be suitable, a depreciation warning can be added to emit when values of the instance (usually `StudyDeck.name`) are referenced after the dialog has been closed to warn add-ons of the need to update, and eventually the blocking option can be removed.

Changes have been tested in MacOS (Apple, PyQt6) and StudyDeck is confirmed to behave normally when called using either approach.

If this approach is successful for StudyDeck, the same changes should be made to other objects that block the main loop.

### CODE REVIEW

**Comment by roxgib** [comment]
For an example of how the use of `QDialog.exec()` affects add-ons, see [here](https://github.com/roxgib/anki-contanki/issues/1).

File: qt/aqt/studydeck.py [hunk 1]
```diff
@@ -36,6 +36,7 @@ def __init__(
         dyn: bool = False,
         buttons: Optional[list[Union[str, QPushButton]]] = None,
         geomKey: str = "default",
+        callback: Callable = None,
```
Comment (dae) [inline qt/aqt/studydeck.py:39]:
```suggestion
        callback: Callable | None = None,
```

File: qt/aqt/studydeck.py [hunk 2]
```diff
@@ -152,6 +157,8 @@ def accept(self) -> None:
             showInfo(tr.decks_please_select_something())
             return
         self.name = self.names[self.form.list.currentRow()]
+        if callable(self.callback):
```
Comment (dae) [inline qt/aqt/studydeck.py:160]: A minor point, but you've tested for truthiness above, so you could probably do so here to be consistent.

**Review by dae — COMMENTED** [review]
Thanks Sam, this looks reasonable, and I appreciate that you've thought about compatibility with existing add-ons. Some minor points below:

**Comment by dae** [comment]
Popping this at the top of the file below the top comments should fix that CI error:

```
from __future__ import annotations
```

**Comment by roxgib** [comment]
I already changed it to Union before I saw you comment, but I'll keep that in mind for the future. 

It's not quite clear to me why the None type hint is required for this line but not e.g. this one...

https://github.com/roxgib/anki/blob/3439483e26511bd036ed681a6c56afaaad1e203c/qt/aqt/studydeck.py#L29

...but I'll read up on type hints more tomorrow

File: qt/aqt/studydeck.py (hunk 2 above)
Comment (roxgib) [reply qt/aqt/studydeck.py]: Updated as suggested

**Comment by dae** [comment]
That line would ideally use a union too, but it wasn't part of your change, and I didn't notice it :-) mypy isn't warning us when we forget it, as we have a bunch of legacy code that will need fixing first. Thanks for the fixes and the contribution!

**Pull request merged**